File: frontend/src/api/tradingSystem.ts
/**
 * 个人交易计划 API
 */
import { ApiClient } from './request'

// ==================== 类型定义 ====================

/** 交易风格 */
export type TradingStyle = 'short_term' | 'medium_term' | 'long_term'

/** 风险偏好 */
export type RiskProfile = 'conservative' | 'balanced' | 'aggressive'

/** 交易计划状态 */
export type TradingSystemStatus = 'draft' | 'published'

/** 选股规则 */
export interface StockSelectionRule {
  analysis_config?: Record<string, any>
  must_have?: Array<Record<string, any>>
  exclude?: Array<Record<string, any>>
  bonus?: Array<Record<string, any>>
}

/** 择时规则 */
export interface TimingRule {
  market_condition?: Record<string, any>
  sector_condition?: Record<string, any>
  entry_signals?: Array<Record<string, any>>
  confirmation?: Array<Record<string, any>>
}

/** 仓位规则 */
export interface PositionRule {
  total_position?: Record<string, number>
  max_per_stock?: number
  max_holdings?: number
  min_holdings?: number
  scaling?: Record<string, any>
}

/** 持仓规则 */
export interface HoldingRule {
  review_frequency?: string
  add_conditions?: Array<Record<string, any>>
  reduce_conditions?: Array<Record<string, any>>
  switch_conditions?: Array<Record<string, any>>
}

/** 风险管理规则 */
export interface RiskManagementRule {
  stop_loss?: Record<string, any>
  take_profit?: Record<string, any>
  time_stop?: Record<string, any>
  logical_stop?: Record<string, any>
}

/** 复盘规则 */
export interface ReviewRule {
  frequency?: string
  checklist?: string[]
  case_save?: Record<string, any>
}

/** 纪律规则 */
export interface DisciplineRule {
  must_not?: Array<Record<string, any>>
  must_do?: Array<Record<string, any>>
  violation_actions?: Array<Record<string, any>>
}

/** 交易计划 */
export interface TradingSystem {
  id?: string
  user_id: string
  name: string
  description?: string
  style: TradingStyle
  risk_profile: RiskProfile
  version: string
  status: TradingSystemStatus
  is_active: boolean

  stock_selection?: StockSelectionRule
  timing?: TimingRule
  position?: PositionRule
  holding?: HoldingRule
  risk_management?: RiskManagementRule
  review?: ReviewRule
  discipline?: DisciplineRule

  // 草稿数据（仅在已发布版本中存在）
  draft_data?: TradingSystemUpdatePayload | null
  draft_updated_at?: string | null

  created_at: string
  updated_at: string
}

/** 创建交易计划请求 */
export interface TradingSystemCreatePayload {
  name: string
  description?: string
  style?: TradingStyle
  risk_profile?: RiskProfile
  
  stock_selection?: StockSelectionRule
  timing?: TimingRule
  position?: PositionRule
  holding?: HoldingRule
  risk_management?: RiskManagementRule
  review?: ReviewRule
  discipline?: DisciplineRule
}

/** 更新交易计划请求 */
export interface TradingSystemUpdatePayload {
  name?: string
  description?: string
  style?: TradingStyle
  risk_profile?: RiskProfile
  is_active?: boolean
  
  stock_selection?: StockSelectionRule
  timing?: TimingRule
  position?: PositionRule
  holding?: HoldingRule
  risk_management?: RiskManagementRule
  review?: ReviewRule
  discipline?: DisciplineRule
}

/** API 响应 */
export interface ApiResponse<T = any> {
  success: boolean
  data: T
  message: string
}

// ==================== API 方法 ====================

/**
 * 创建交易计划
 */
export function createTradingSystem(payload: TradingSystemCreatePayload) {
  return ApiClient.post<{ system: TradingSystem }>('/api/v1/trading-systems', payload)
}

/**
 * 获取交易计划列表
 */
export function getTradingSystems(isActive?: boolean) {
  const params = isActive !== undefined ? { is_active: isActive } : {}
  return ApiClient.get<{ systems: TradingSystem[]; total: number }>('/api/v1/trading-systems', params)
}

/**
 * 获取激活的交易计划
 */
export function getActiveTradingSystem() {
  return ApiClient.get<{ system: TradingSystem | null }>('/api/v1/trading-systems/active')
}

/**
 * 获取交易计划详情
 */
export function getTradingSystem(systemId: string) {
  return ApiClient.get<{ system: TradingSystem }>(`/api/v1/trading-systems/${systemId}`)
}

/**
 * 更新交易计划
 */
export function updateTradingSystem(
  systemId: string,
  payload: TradingSystemUpdatePayload,
  saveAsDraft: boolean = false
) {
  const params = saveAsDraft ? { save_as_draft: true } : {}
  return ApiClient.put<{ system: TradingSystem }>(
    `/api/v1/trading-systems/${systemId}`,
    payload,
    { params }
  )
}

/**
 * 删除交易计划
 */
export function deleteTradingSystem(systemId: string) {
  return ApiClient.delete<void>(`/api/v1/trading-systems/${systemId}`)
}

/**
 * 激活交易计划
 */
export function activateTradingSystem(systemId: string) {
  return ApiClient.post<{ system: TradingSystem }>(`/api/v1/trading-systems/${systemId}/activate`, {})
}

/** 交易计划评估结果 */
export interface TradingPlanEvaluation {
  overall_score: number
  strengths: string[]
  weaknesses: string[]
  suggestions: string[]
  detailed_analysis: string
  evaluation_date: string
}

/**
 * AI评估交易计划（已保存的计划）
 */
export function evaluateTradingSystem(systemId: string) {
  return ApiClient.post<{ evaluation: TradingPlanEvaluation }>(`/api/v1/trading-systems/${systemId}/evaluate`, {})
}

/**
 * AI评估交易计划草稿（未保存的计划）
 */
export function evaluateTradingPlanDraft(payload: TradingSystemCreatePayload) {
  return ApiClient.post<{ evaluation: TradingPlanEvaluation }>('/api/v1/trading-systems/evaluate-draft', payload)
}

/**
 * 评估历史记录项
 */
export interface EvaluationHistoryItem {
  evaluation_id: string
  system_id: string
  system_name: string
  system_version: string
  overall_score: number
  grade: string
  created_at: string
  summary: string
}

/**
 * 评估历史列表响应
 */
export interface EvaluationHistoryResponse {
  items: EvaluationHistoryItem[]
  total: number
  page: number
  page_size: number
}

/**
 * 获取交易计划评估历史
 */
export function getEvaluationHistory(systemId: string, page: number = 1, pageSize: number = 10) {
  return ApiClient.get<EvaluationHistoryResponse>(`/api/v1/trading-systems/${systemId}/evaluations`, {
    params: { page, page_size: pageSize }
  })
}

/**
 * 获取评估详情
 */
export function getEvaluationDetail(evaluationId: string) {
  return ApiClient.get<{ evaluation: any }>(`/api/v1/trading-systems/evaluations/${evaluationId}`)
}

// ==================== 版本管理 ====================

/** 交易计划版本 */
export interface TradingSystemVersion {
  id?: string
  system_id: string
  version: string
  improvement_summary: string
  snapshot: TradingSystem
  created_at: string
  created_by: string
}

/** 创建版本请求 */
export interface TradingSystemVersionCreatePayload {
  improvement_summary: string
  new_version?: string
}

/**
 * 创建交易计划新版本
 */
export function createTradingSystemVersion(systemId: string, payload: TradingSystemVersionCreatePayload) {
  return ApiClient.post<{ version: TradingSystemVersion }>(`/api/v1/trading-systems/${systemId}/versions`, payload)
}

/**
 * 获取交易计划的所有版本
 */
export function getTradingSystemVersions(systemId: string) {
  return ApiClient.get<{ versions: TradingSystemVersion[]; total: number }>(`/api/v1/trading-systems/${systemId}/versions`)
}

/**
 * 获取版本详情
 */
export function getTradingSystemVersion(versionId: string) {
  return ApiClient.get<{ version: TradingSystemVersion }>(`/api/v1/trading-systems/versions/${versionId}`)
}

/**
 * 发布交易计划（创建新版本并更新状态为已发布）
 */
export interface TradingSystemPublishPayload {
  improvement_summary: string
  new_version?: string
  update_data?: TradingSystemUpdatePayload
}

export function publishTradingSystem(
  systemId: string,
  publishPayload: TradingSystemPublishPayload
) {
  return ApiClient.post<{ system: TradingSystem }>(
    `/api/v1/trading-systems/${systemId}/publish`,
    publishPayload
  )
}

File: app/models/trading_system.py
"""
个人交易计划数据模型
"""

from datetime import datetime
from typing import Optional, List, Dict, Any
from pydantic import BaseModel, Field, model_validator
from enum import Enum


class TradingStyle(str, Enum):
    """交易风格"""
    SHORT_TERM = "short_term"      # 短线（1天-1周）
    MEDIUM_TERM = "medium_term"    # 中线（1周-1月）
    LONG_TERM = "long_term"        # 长线（1月-1年+）


class RiskProfile(str, Enum):
    """风险偏好"""
    CONSERVATIVE = "conservative"  # 保守
    BALANCED = "balanced"          # 稳健
    AGGRESSIVE = "aggressive"      # 激进


class TradingSystemStatus(str, Enum):
    """交易计划状态"""
    DRAFT = "draft"        # 草稿
    PUBLISHED = "published"  # 已发布


class StockSelectionRule(BaseModel):
    """选股规则模型"""
    analysis_config: Dict[str, Any] = Field(default_factory=dict, description="分析配置")
    must_have: List[Dict[str, Any]] = Field(default_factory=list, description="必备条件")
    exclude: List[Dict[str, Any]] = Field(default_factory=list, description="排除条件")
    bonus: List[Dict[str, Any]] = Field(default_factory=list, description="加分条件")


class TimingRule(BaseModel):
    """择时规则模型"""
    market_condition: Dict[str, Any] = Field(default_factory=dict, description="大盘条件")
    sector_condition: Dict[str, Any] = Field(default_factory=dict, description="行业条件")
    entry_signals: List[Dict[str, Any]] = Field(default_factory=list, description="买入信号")
    confirmation: List[Dict[str, Any]] = Field(default_factory=list, description="确认条件")


class PositionRule(BaseModel):
    """仓位规则模型"""
    total_position: Dict[str, float] = Field(
        default_factory=lambda: {"bull": 0.8, "range": 0.5, "bear": 0.2},
        description="总仓位控制（按市况）"
    )
    max_per_stock: float = Field(default=0.3, description="单只股票上限")
    max_holdings: int = Field(default=5, description="最大持股数量")
    min_holdings: int = Field(default=1, description="最小持股数量")
    scaling: Dict[str, Any] = Field(default_factory=dict, description="分批策略")


class HoldingRule(BaseModel):
    """持仓规则模型"""
    review_frequency: str = Field(default="daily", description="检视频率")
    add_conditions: List[Dict[str, Any]] = Field(default_factory=list, description="加仓条件")
    reduce_conditions: List[Dict[str, Any]] = Field(default_factory=list, description="减仓条件")
    switch_conditions: List[Dict[str, Any]] = Field(default_factory=list, description="换股条件")


class RiskManagementRule(BaseModel):
    """风险管理规则模型"""
    stop_loss: Dict[str, Any] = Field(default_factory=dict, description="止损设置")
    take_profit: Dict[str, Any] = Field(default_factory=dict, description="止盈设置")
    time_stop: Dict[str, Any] = Field(default_factory=dict, description="时间止损")
    logical_stop: Dict[str, Any] = Field(default_factory=dict, description="逻辑止损")

    @staticmethod
    def _sanitize_stop_loss(stop_loss: Optional[Dict[str, Any]]) -> Dict[str, Any]:
        source = stop_loss or {}
        stop_type = source.get("type", "percentage")

        if stop_type == "technical":
            return {
                "type": "technical",
                "description": source.get("description", "")
            }

        if stop_type == "atr":
            return {
                "type": "atr",
                "atr_multiplier": float(source.get("atr_multiplier", 2.0))
            }

        return {
            "type": "percentage",
            "percentage": float(source.get("percentage", 0.08))
        }

    @staticmethod
    def _sanitize_take_profit(take_profit: Optional[Dict[str, Any]]) -> Dict[str, Any]:
        source = take_profit or {}
        take_type = source.get("type", "percentage")

        if take_type == "trailing":
            return {
                "type": "trailing",
                "trailing_pullback_pct": float(source.get("trailing_pullback_pct", 0.08)),
                "activation_profit_pct": float(source.get("activation_profit_pct", 0.1)),
                "reference": source.get("reference", "highest_price")
            }

        if take_type == "scaled":
            levels = source.get("levels") if isinstance(source.get("levels"), list) else []
            if not levels:
                levels = [
                    {"target_profit_pct": 0.2, "sell_ratio": 0.3},
                    {"target_profit_pct": 0.35, "sell_ratio": 0.3},
                    {"target_profit_pct": 0.5, "sell_ratio": 0.4},
                ]
            return {
                "type": "scaled",
                "levels": levels
            }

        return {
            "type": "percentage",
            "percentage": float(source.get("percentage", 0.2))
        }

    @model_validator(mode="after")
    def normalize_exclusive_fields(self):
        self.stop_loss = self._sanitize_stop_loss(self.stop_loss)
        self.take_profit = self._sanitize_take_profit(self.take_profit)
        return self


class ReviewRule(BaseModel):
    """复盘规则模型"""
    frequency: str = Field(default="weekly", description="复盘频率")
    checklist: List[str] = Field(default_factory=list, description="复盘内容")
    case_save: Dict[str, Any] = Field(default_factory=dict, description="案例保存规则")


class DisciplineRule(BaseModel):
    """纪律规则模型"""
    must_not: List[Dict[str, Any]] = Field(default_factory=list, description="绝对禁止")
    must_do: List[Dict[str, Any]] = Field(default_factory=list, description="必须做到")
    violation_actions: List[Dict[str, Any]] = Field(default_factory=list, description="违规处理")


class TradingSystem(BaseModel):
    """交易计划主模型"""
    id: Optional[str] = Field(None, description="系统ID")
    user_id: str = Field(..., description="所属用户ID")
    name: str = Field(..., description="系统名称")
    description: str = Field(default="", description="系统描述")
    style: TradingStyle = Field(default=TradingStyle.MEDIUM_TERM, description="交易风格")
    risk_profile: RiskProfile = Field(default=RiskProfile.BALANCED, description="风险偏好")
    version: str = Field(default="1.0.0", description="版本号")
    status: TradingSystemStatus = Field(default=TradingSystemStatus.DRAFT, description="状态：草稿/已发布")
    is_active: bool = Field(default=True, description="是否激活")

    # 六大模块规则（正式版本）
    stock_selection: StockSelectionRule = Field(default_factory=StockSelectionRule, description="选股规则")
    timing: TimingRule = Field(default_factory=TimingRule, description="择时规则")
    position: PositionRule = Field(default_factory=PositionRule, description="仓位规则")
    holding: HoldingRule = Field(default_factory=HoldingRule, description="持仓规则")
    risk_management: RiskManagementRule = Field(default_factory=RiskManagementRule, description="风险管理规则")
    review: ReviewRule = Field(default_factory=ReviewRule, description="复盘规则")
    discipline: DisciplineRule = Field(default_factory=DisciplineRule, description="纪律规则")

    # 草稿数据（仅在已发布版本中存在，保存草稿时使用）
    draft_data: Optional[Dict[str, Any]] = Field(None, description="草稿数据，包含修改后的规则")
    draft_updated_at: Optional[datetime] = Field(None, description="草稿更新时间")

    # 元数据
    created_at: datetime = Field(default_factory=datetime.utcnow, description="创建时间")
    updated_at: datetime = Field(default_factory=datetime.utcnow, description="更新时间")

    class Config:
        populate_by_name = True
        json_encoders = {
            datetime: lambda v: v.isoformat()
        }


class TradingSystemCreate(BaseModel):
    """创建交易计划请求模型"""
    name: str = Field(..., description="系统名称")
    description: str = Field(default="", description="系统描述")
    style: TradingStyle = Field(default=TradingStyle.MEDIUM_TERM, description="交易风格")
    risk_profile: RiskProfile = Field(default=RiskProfile.BALANCED, description="风险偏好")
    
    stock_selection: Optional[StockSelectionRule] = None
    timing: Optional[TimingRule] = None
    position: Optional[PositionRule] = None
    holding: Optional[HoldingRule] = None
    risk_management: Optional[RiskManagementRule] = None
    review: Optional[ReviewRule] = None
    discipline: Optional[DisciplineRule] = None


class TradingSystemUpdate(BaseModel):
    """更新交易计划请求模型"""
    name: Optional[str] = None
    description: Optional[str] = None
    style: Optional[TradingStyle] = None
    risk_profile: Optional[RiskProfile] = None
    status: Optional[TradingSystemStatus] = None
    is_active: Optional[bool] = None
    
    stock_selection: Optional[StockSelectionRule] = None
    timing: Optional[TimingRule] = None
    position: Optional[PositionRule] = None
    holding: Optional[HoldingRule] = None
    risk_management: Optional[RiskManagementRule] = None
    review: Optional[ReviewRule] = None
    discipline: Optional[DisciplineRule] = None


class TradingSystemPublish(BaseModel):
    """发布交易计划请求模型"""
    improvement_summary: str = Field(..., description="改进总结（发布时必填）")
    new_version: Optional[str] = Field(None, description="新版本号，如果不提供则自动递增")
    # 可选的更新数据（如果有修改）
    update_data: Optional[TradingSystemUpdate] = Field(None, description="可选的更新数据")


class TradingSystemVersion(BaseModel):
    """交易计划版本模型"""
    id: Optional[str] = Field(None, description="版本ID")
    system_id: str = Field(..., description="关联的交易计划ID")
    version: str = Field(..., description="版本号，如 '1.0.0', '2.0.0'")
    improvement_summary: str = Field(default="", description="改进总结")
    
    # 完整的系统快照
    snapshot: TradingSystem = Field(..., description="该版本的完整系统快照")
    
    # 元数据
    created_at: datetime = Field(default_factory=datetime.utcnow, description="创建时间")
    created_by: str = Field(..., description="创建者用户ID")
    
    class Config:
        populate_by_name = True
        json_encoders = {
            datetime: lambda v: v.isoformat()
        }


class TradingSystemVersionCreate(BaseModel):
    """创建交易计划版本请求模型"""
    improvement_summary: str = Field(..., description="改进总结")
    new_version: Optional[str] = Field(None, description="新版本号，如果不提供则自动递增")

